### PR TITLE
Abstract LibGC usages. Enable none garbage collector.

### PR DESCRIPTION
**Pull request by ysbaddaden**
All direct calls do `LibGC` are now abstracted under the `GC` module. The null and boehm garbage collectors were updated to follow the new API (the null GC compiles again). The `__crystal_malloc` internal functions were also abstracted to call into `GC.malloc` where implementation / bindings are expected to be.

Also introduces a `GC_NULL` compilation flag to select the null garbage collection instead of Boehm GC. 

This patch allows experimenting with alternative or custom garbage collectors.

### CODE REVIEW

**Comment by ysbaddaden** [comment]
(now nobody will guess whatever I've been getting my hands dirty with these past few weeks)

File: src/gc.cr [hunk 1]
```diff
@@ -1,13 +1,76 @@
...
+    malloc_atomic(LibC::SizeT.new(size))
   end
 
   def self.realloc(pointer : Void*, size : Int)
-    __crystal_realloc(pointer, size.to_u32)
+    realloc(pointer, LibC::SizeT.new(size))
   end
 end
+
+{% if flag?(:GC_NULL)%}
```
Comment (RX14) [inline src/gc.cr]: Typically flags so far have been lowercase so far (`without_openssl`, etc), why not `null_gc`?

Comment (RX14) [reply src/gc.cr]: Or maybe `-Dno_gc`?

Comment (RX14) [reply src/gc.cr]: If that's the case, maybe we should rename `gc/null` to `gc/none`?

File: src/gc.cr [hunk 2]
```diff
@@ -1,13 +1,76 @@
+# :nodoc:
+fun __crystal_malloc(size : UInt32) : Void*
+  GC.malloc(LibC::SizeT.new(size))
+end
+
+# :nodoc:
+fun __crystal_malloc_atomic(size : UInt32) : Void*
+  GC.malloc(LibC::SizeT.new(size))
```
Comment (bew) [inline src/gc.cr]: What's the point if `__crystal_malloc_atomic` & `__crystal_malloc` have the same implementation?
It doesn't look right.

File: src/gc.cr [hunk 3]
```diff
@@ -1,13 +1,76 @@
...
+
+# :nodoc:
+fun __crystal_malloc_atomic64(size : UInt64) : Void*
+  {% if flag?(:bits32) %}
+    if size > UInt32::MAX
+      raise ArgumentError.new("Given size is bigger than UInt32::MAX")
+    end
+  {% end %}
+
+  GC.malloc(LibC::SizeT.new(size))
```
Comment (bcardiff) [inline src/gc.cr]: Ditto regarding atomic

File: src/gc.cr (hunk 1 above)
Comment (bcardiff) [reply src/gc.cr]: I wouldn’t use “no gc”. Seems more like a toggle to switch to a c++ like memory management to me.

Comment (ysbaddaden) [reply src/gc.cr]: Habit? Too much C reading? Constant-like looking? Anyway, can change to whatever.

**Comment by ysbaddaden** [comment]
Fixed thread issue in specs. Fixed atomic malloc calls. Renamed `GC_NULL` flag to `gc_none`.

**Comment by RX14** [comment]
@ysbaddaden please make `fixup!` commits so I can review just the changes!

File: src/gc.cr [hunk 4]
```diff
@@ -1,13 +1,76 @@
...
+    malloc_atomic(LibC::SizeT.new(size))
   end
 
   def self.realloc(pointer : Void*, size : Int)
-    __crystal_realloc(pointer, size.to_u32)
+    realloc(pointer, LibC::SizeT.new(size))
   end
 end
+
+{% if flag?(:gc_none)%}
```
Comment (Sija) [inline src/gc.cr]: missing space before `%}`

File: src/gc.cr (hunk 3 above)
Comment (luislavena) [reply src/gc.cr]: @bcardiff I think this was addressed in latest change, can you check again? Thank you.

**Pull request merged**